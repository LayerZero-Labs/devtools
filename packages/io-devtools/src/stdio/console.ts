/**
 * Executes the callback while mocking the console methods.
 * This will cause all the console output to be silenced.
 *
 * This method will return a promise resolved with the
 * resolved/returned value of the callback:
 *
 * ```
 * const result = await withMutedConsole(() => 7) // result === 7
 * const result = await withMutedConsole(async () => 7) // result === 7
 * ```
 *
 * @param {() => T | Promise<T>} callback
 */
export const withMutedConsole = async <T>(callback: () => T | Promise<T>): Promise<Awaited<T>> => {
    const originalLog = console.log
    const originalInfo = console.info
    const originalWarn = console.warn
    const originalError = console.error
    const originalDebug = console.debug
    const noop = () => {}

    try {
        console.log = noop
        console.info = noop
        console.warn = noop
        console.error = noop
        console.debug = noop

        return await callback()
    } finally {
        console.log = originalLog
        console.info = originalInfo
        console.warn = originalWarn
        console.error = originalError
        console.debug = originalDebug
    }
}
