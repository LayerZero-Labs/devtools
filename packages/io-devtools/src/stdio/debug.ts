export { debuglog, type DebugLogger } from 'node:util'
