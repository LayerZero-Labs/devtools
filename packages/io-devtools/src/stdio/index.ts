export * from './logger'
export * from './printer'
export * from './prompts'
