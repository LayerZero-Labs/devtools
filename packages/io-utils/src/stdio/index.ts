export * from './printer'
export * from './prompts'
