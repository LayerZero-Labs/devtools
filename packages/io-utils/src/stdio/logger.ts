import { createLogger as createWinstonLogger, format, transports, Logger } from 'winston'

/**
 * Valid logging levels
 */
export enum LogLevel {
    error = 'error',
    warn = 'warn',
    info = 'info',
    http = 'http',
    verbose = 'verbose',
    debug = 'debug',
    silly = 'silly',
}

/**
 * Type assertion utility for LogLevel
 *
 * @param value `unknown`
 * @returns `value is LogLevel`
 */
const isLogLevel = (value: unknown): value is LogLevel => typeof value === 'string' && value in LogLevel

let DEFAULT_LOG_LEVEL = LogLevel.info

/**
 * Sets the default log level used when creating new loggers.
 *
 * @param level `LogLevel`
 */
export const setDefaultLogLevel = (level: string) => {
    if (!isLogLevel(level)) {
        console.warn(
            `Invalid log level specified: ${level}. Ignoring and keeping the current value of ${DEFAULT_LOG_LEVEL}`
        )

        return
    }

    DEFAULT_LOG_LEVEL = level
}

/**
 * Creates a general-purpose logger
 *
 * @param level `LogLevel` Default to the globally set log level (@see `setDefaultLogLevel`)
 * @param logFormat `LogFormat` Optional Winston LogForm.Format instance
 *
 * @returns `Logger`
 */
export const createLogger = (level: string = DEFAULT_LOG_LEVEL, logFormat = format.cli()) =>
    createWinstonLogger({
        level,
        format: logFormat,
        transports: [new transports.Console()],
    })

/**
 * Creates a logger a module specific logging.
 *
 * The messages will be prefixed with the module name:
 *
 * `[module] Some message`
 *
 * @param {string} module
 * @param {string} [level] Default to the globally set log level (@see `setDefaultLogLevel`)
 *
 * @returns {Logger}
 */
export const createModuleLogger = (module: string, level: string = DEFAULT_LOG_LEVEL): Logger =>
    createLogger(level, format.combine(prefix({ label: module }), format.cli()))

/**
 * Creates a logger for network-to-network interactions.
 *
 * The messages will be prefixed with the network pair:
 *
 * `[ethereum-mainnet ➝ avalanche-mainnet] Some message`
 *
 * @param {string} sourceModule
 * @param {string} destinationModule
 * @param {string} [level] Default to the globally set log level (@see `setDefaultLogLevel`)
 *
 * @returns {Logger}
 */
export const createModuleInteractionLogger = (
    sourceModule: string,
    destinationModule: string,
    level: string = DEFAULT_LOG_LEVEL
): Logger =>
    createLogger(level, format.combine(prefix({ label: `${sourceModule} ➝ ${destinationModule}` }), format.cli()))

/**
 * Helper utility that prefixes logged messages
 * with label (wrapped in square brackets)
 *
 * An example of logged message with `ethereum-mainnet` label:
 *
 * `[ethereum-mainnet] Some message`
 */
const prefix = format((info, { label }) => ({
    ...info,
    message: `${label ? `[${label}] ` : ''}${info.message}`,
}))
