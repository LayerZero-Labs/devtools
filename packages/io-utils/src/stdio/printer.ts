import Table, { HorizontalTableRow } from 'cli-table3'
import type { ZodError } from 'zod'

export const printJson = (obj: unknown, pretty = true): string =>
    JSON.stringify(
        obj,
        (key, value) => (typeof value === 'bigint' ? value.toString(10) : value),
        pretty ? '\t' : undefined
    )

export const printRecord = <TRecord extends object>(obj: TRecord, title?: string | number): string => {
    const table = new Table({
        wordWrap: true,
        wrapOnWordBoundary: false,
    })

    const headers: HorizontalTableRow[] = title == null ? [] : [[{ content: title, colSpan: 2 }]]
    const rows = Object.entries(obj).map(([key, value]): HorizontalTableRow => {
        switch (true) {
            case value == null:
            case value instanceof Date:
            case typeof value !== 'object':
                return [key, String(value)]

            default:
                return [key, printRecord(value)]
        }
    })

    return table.push(...headers, ...rows), table.toString()
}

/**
 * Helper utility for printing out boolean values
 *
 * @param {boolean | null | undefined} value
 * @returns {string}
 */
export const printBoolean = (value: boolean | null | undefined): string => (value == null ? '∅' : value ? '✅' : '❌')

export const printZodErrors = (error: ZodError<unknown>): string => {
    // Here we will go through all the errors and prefix them with the name
    // of the property on which they happened, if any
    const errors = error.flatten((issue) => {
        const propertyPath = issue.path?.join('.') ?? ''
        if (propertyPath === '') return issue.message

        return `Property '${propertyPath}': ${issue.message}`
    })

    // These are the errors coming from a mismatch on the root object - e.g. when a whole object is completely missing
    const formErrors = errors.formErrors

    // These are errors coming from object properties
    const fieldErrors = Object.values<unknown[]>(errors.fieldErrors).flat()

    // Now we take all the errors we got
    const allErrors = [...formErrors, ...fieldErrors]

    // And concatenate
    return allErrors.join(`\n`)
}
