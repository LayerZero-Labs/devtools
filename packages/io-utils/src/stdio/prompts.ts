import assert from 'assert'
import prompts from 'prompts'

export const promptToContinue = async (
    message: string = 'Do you want to continue?',
    defaultValue = true
): Promise<boolean> => {
    const { value } = await prompts({
        type: 'confirm',
        name: 'value',
        message,
        initial: defaultValue,
    })

    assert(typeof value === 'boolean', `Invariant error: Expected a boolean response, got ${value}`)

    return value
}
