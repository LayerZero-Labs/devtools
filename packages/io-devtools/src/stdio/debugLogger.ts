import { createLogger } from './logger'

/**
 * Provides simple console logging helpers used across devtools examples.
 */
export class DebugLogger {
    static keyValue(key: string, value: any, indentLevel = 0) {
        const indent = ' '.repeat(indentLevel * 2)
        console.log(`${indent}\x1b[33m${key}:\x1b[0m ${value}`)
    }

    static keyHeader(key: string, indentLevel = 0) {
        const indent = ' '.repeat(indentLevel * 2)
        console.log(`${indent}\x1b[33m${key}:\x1b[0m`)
    }

    static header(text: string) {
        console.log(`\x1b[36m${text}\x1b[0m`)
    }

    static separator() {
        console.log('\x1b[90m----------------------------------------\x1b[0m')
    }

    static printErrorAndFixSuggestion(type: KnownErrors, errorMsg?: string) {
        const fixInfo = ERRORS_FIXES_MAP[type]
        if (!fixInfo) {
            console.log(`\x1b[31mError:\x1b[0m Unknown error type "${type}"`)
            return
        }

        const errorOutput = errorMsg ? `${type}: (${errorMsg})` : type

        console.log(`\x1b[31mError:\x1b[0m ${errorOutput}`)
        console.log(`\x1b[32mFix suggestion:\x1b[0m ${fixInfo.tip}`)
        if (fixInfo.info) {
            console.log(`\x1b[34mElaboration:\x1b[0m ${fixInfo.info}`)
        }
        console.log()
    }

    static printWarning(type: KnownWarnings, message?: string) {
        const label = `\x1b[33mWarning:\x1b[0m`
        console.log(`${label} ${type}${message ? ` â€“ ${message}` : ''}`)
    }

    static printLayerZeroOutput(type: KnownOutputs, payload?: string) {
        logger.info(`${payload ? ' ' + payload : ''}`)
    }
}

export enum KnownErrors {
    ULN_INIT_CONFIG_SKIPPED = 'ULN_INIT_CONFIG_SKIPPED',
    ERROR_QUOTING_NATIVE_GAS_COST = 'ERROR_QUOTING_NATIVE_GAS_COST',
    ERROR_SENDING_TRANSACTION = 'ERROR_SENDING_TRANSACTION',
    ERROR_GETTING_HRE = 'ERROR_GETTING_HARDHAT_RUNTIME_ENVIRONMENT_FOR_NETWORK',
}

export enum KnownWarnings {
    OFT_PROGRAM_NOT_DEPLOYED = 'OFT Program Not Deployed',
    USING_OVERRIDE_OFT = 'Using address provided as an OFT deployment',
    SOLANA_DEPLOYMENT_MISSING_OFT_STORE = 'Solana deployment missing OFT store',
    SOLANA_DEPLOYMENT_NOT_FOUND = 'SOLANA_DEPLOYMENT_NOT_FOUND',
    ERROR_LOADING_SOLANA_DEPLOYMENT = 'Error loading local Solana deployment',
}

export enum KnownOutputs {
    TX_HASH = 'Transaction hash',
    EXPLORER_LINK = 'LayerZero scan link',
    SENT_VIA_OFT = 'OFT sent successfully',
}

interface ErrorFixInfo {
    tip: string
    info?: string
}

export const ERRORS_FIXES_MAP: Record<KnownErrors, ErrorFixInfo> = {
    [KnownErrors.ULN_INIT_CONFIG_SKIPPED]: {
        tip: 'Did you run `npx hardhat lz:oft:solana:init-config --oapp-config <LZ_CONFIG_FILE_NAME> ?',
        info: 'You must run lz:oft:solana:init-config once before you run lz:oapp:wire. If you have added new pathways, you must also run lz:oft:solana:init-config again.',
    },
    [KnownErrors.ERROR_QUOTING_NATIVE_GAS_COST]: {
        tip: 'Have you run `npx hardhat lz:oapp:config:get --oapp-config <LZ_CONFIG_FILE_NAME>` and checked that you correctly configured the pathway?',
        info: 'LayerZero pathways require that a default Endpoint, Message Library, and DVN configuration exists for messaging to work. See https://layerzeroscan.com/tools/defaults for more information.',
    },
    [KnownErrors.ERROR_SENDING_TRANSACTION]: {
        tip: 'Have you correctly passed the quoteSend() result to the send() function?',
        info: 'To quote the native gas cost needed to send a message, you must pass the result of quoteSend() to the send() function.',
    },
    [KnownErrors.ERROR_GETTING_HRE]: {
        tip: 'Have you added the srcEid network to your `./hardhat.config.ts` file?',
        info: 'If you loaded a custom OFT deployment from an EVM network, you must add the deployment srcEid to your `./hardhat.config.ts` file for the OFT to be found.',
    },
}

export const WARNINGS_FIXES_MAP: Record<KnownWarnings, ErrorFixInfo> = {
    [KnownWarnings.SOLANA_DEPLOYMENT_NOT_FOUND]: {
        tip: 'Did you run `npx hardhat lz:oft:solana:create` ?',
        info: 'The Solana deployment file is required to run config tasks. The default path is ./deployments/solana-<mainnet/testnet>/OFT.json',
    },
    [KnownWarnings.OFT_PROGRAM_NOT_DEPLOYED]: {
        tip: 'Deploy the OFT program first',
        info: 'The OFT program must be deployed before proceeding with other operations',
    },
    [KnownWarnings.USING_OVERRIDE_OFT]: {
        tip: 'Using external OFT deployment',
        info: 'This is expected when using an external OFT deployment',
    },
    [KnownWarnings.SOLANA_DEPLOYMENT_MISSING_OFT_STORE]: {
        tip: 'OFT store is missing from deployment',
        info: 'The OFT store must be initialized in the deployment',
    },
    [KnownWarnings.ERROR_LOADING_SOLANA_DEPLOYMENT]: {
        tip: 'Failed to load Solana deployment',
        info: 'Check if the deployment file exists and is properly formatted',
    },
}

const logger = createLogger()
