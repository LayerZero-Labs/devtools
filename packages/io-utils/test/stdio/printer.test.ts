import { printRecord } from '@/stdio'

describe('stdio/printer', () => {
    let CI: string | undefined

    // We want to force non-colorful output to make the snapshots more readable
    // and to make the tests consistent both in & out of the container
    beforeAll(() => {
        CI = process.env.CI

        process.env.CI = '1'
    })

    afterAll(() => {
        process.env.CI = CI
    })

    describe('printRecord', () => {
        const record = {
            string: 'something',
            number: 0,
            boolean: false,
            bigint: BigInt(1),
            null: null,
            undefined: undefined,
            symbol: Symbol('oh hello'),
            function: () => {},
        }
        const list = Object.values(record)

        it('should return an empty string for an empty array', () => {
            expect(printRecord([])).toMatchSnapshot()
        })

        it('should return a list of items with indices for an array', () => {
            expect(printRecord(list)).toMatchSnapshot()
        })

        it('should return a list of items with indices with a title for an array', () => {
            expect(printRecord(list, 'My items')).toMatchSnapshot()
        })

        it('should return a list of items with indices for an object', () => {
            expect(printRecord(record)).toMatchSnapshot()
        })

        it('should return a list of items with indices with a title for an object', () => {
            expect(printRecord(record, 'My items')).toMatchSnapshot()
        })

        it('should return a nested list for a nested object', () => {
            expect(
                printRecord({
                    ...record,
                    nestedArray: list,
                    nestedObject: record,
                })
            ).toMatchSnapshot()
        })

        it('should return a nested list for a nested object in array', () => {
            expect(printRecord([...list, list, record])).toMatchSnapshot()
        })
    })
})
