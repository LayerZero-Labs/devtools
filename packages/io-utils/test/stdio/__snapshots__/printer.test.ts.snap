// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stdio/printer printRecord should return a list of items with indices for an array 1`] = `
"┌───┬──────────────────┐
│ 0 │ something        │
├───┼──────────────────┤
│ 1 │ 0                │
├───┼──────────────────┤
│ 2 │ false            │
├───┼──────────────────┤
│ 3 │ 1                │
├───┼──────────────────┤
│ 4 │ null             │
├───┼──────────────────┤
│ 5 │ undefined        │
├───┼──────────────────┤
│ 6 │ Symbol(oh hello) │
├───┼──────────────────┤
│ 7 │ () => { }        │
└───┴──────────────────┘"
`;

exports[`stdio/printer printRecord should return a list of items with indices for an object 1`] = `
"┌───────────┬──────────────────┐
│ string    │ something        │
├───────────┼──────────────────┤
│ number    │ 0                │
├───────────┼──────────────────┤
│ boolean   │ false            │
├───────────┼──────────────────┤
│ bigint    │ 1                │
├───────────┼──────────────────┤
│ null      │ null             │
├───────────┼──────────────────┤
│ undefined │ undefined        │
├───────────┼──────────────────┤
│ symbol    │ Symbol(oh hello) │
├───────────┼──────────────────┤
│ function  │ () => { }        │
└───────────┴──────────────────┘"
`;

exports[`stdio/printer printRecord should return a list of items with indices with a title for an array 1`] = `
"┌──────────────────────┐
│ My items             │
├───┬──────────────────┤
│ 0 │ something        │
├───┼──────────────────┤
│ 1 │ 0                │
├───┼──────────────────┤
│ 2 │ false            │
├───┼──────────────────┤
│ 3 │ 1                │
├───┼──────────────────┤
│ 4 │ null             │
├───┼──────────────────┤
│ 5 │ undefined        │
├───┼──────────────────┤
│ 6 │ Symbol(oh hello) │
├───┼──────────────────┤
│ 7 │ () => { }        │
└───┴──────────────────┘"
`;

exports[`stdio/printer printRecord should return a list of items with indices with a title for an object 1`] = `
"┌──────────────────────────────┐
│ My items                     │
├───────────┬──────────────────┤
│ string    │ something        │
├───────────┼──────────────────┤
│ number    │ 0                │
├───────────┼──────────────────┤
│ boolean   │ false            │
├───────────┼──────────────────┤
│ bigint    │ 1                │
├───────────┼──────────────────┤
│ null      │ null             │
├───────────┼──────────────────┤
│ undefined │ undefined        │
├───────────┼──────────────────┤
│ symbol    │ Symbol(oh hello) │
├───────────┼──────────────────┤
│ function  │ () => { }        │
└───────────┴──────────────────┘"
`;

exports[`stdio/printer printRecord should return a nested list for a nested object 1`] = `
"┌──────────────┬──────────────────────────────────┐
│ string       │ something                        │
├──────────────┼──────────────────────────────────┤
│ number       │ 0                                │
├──────────────┼──────────────────────────────────┤
│ boolean      │ false                            │
├──────────────┼──────────────────────────────────┤
│ bigint       │ 1                                │
├──────────────┼──────────────────────────────────┤
│ null         │ null                             │
├──────────────┼──────────────────────────────────┤
│ undefined    │ undefined                        │
├──────────────┼──────────────────────────────────┤
│ symbol       │ Symbol(oh hello)                 │
├──────────────┼──────────────────────────────────┤
│ function     │ () => { }                        │
├──────────────┼──────────────────────────────────┤
│ nestedArray  │ ┌───┬──────────────────┐         │
│              │ │ 0 │ something        │         │
│              │ ├───┼──────────────────┤         │
│              │ │ 1 │ 0                │         │
│              │ ├───┼──────────────────┤         │
│              │ │ 2 │ false            │         │
│              │ ├───┼──────────────────┤         │
│              │ │ 3 │ 1                │         │
│              │ ├───┼──────────────────┤         │
│              │ │ 4 │ null             │         │
│              │ ├───┼──────────────────┤         │
│              │ │ 5 │ undefined        │         │
│              │ ├───┼──────────────────┤         │
│              │ │ 6 │ Symbol(oh hello) │         │
│              │ ├───┼──────────────────┤         │
│              │ │ 7 │ () => { }        │         │
│              │ └───┴──────────────────┘         │
├──────────────┼──────────────────────────────────┤
│ nestedObject │ ┌───────────┬──────────────────┐ │
│              │ │ string    │ something        │ │
│              │ ├───────────┼──────────────────┤ │
│              │ │ number    │ 0                │ │
│              │ ├───────────┼──────────────────┤ │
│              │ │ boolean   │ false            │ │
│              │ ├───────────┼──────────────────┤ │
│              │ │ bigint    │ 1                │ │
│              │ ├───────────┼──────────────────┤ │
│              │ │ null      │ null             │ │
│              │ ├───────────┼──────────────────┤ │
│              │ │ undefined │ undefined        │ │
│              │ ├───────────┼──────────────────┤ │
│              │ │ symbol    │ Symbol(oh hello) │ │
│              │ ├───────────┼──────────────────┤ │
│              │ │ function  │ () => { }        │ │
│              │ └───────────┴──────────────────┘ │
└──────────────┴──────────────────────────────────┘"
`;

exports[`stdio/printer printRecord should return a nested list for a nested object in array 1`] = `
"┌───┬──────────────────────────────────┐
│ 0 │ something                        │
├───┼──────────────────────────────────┤
│ 1 │ 0                                │
├───┼──────────────────────────────────┤
│ 2 │ false                            │
├───┼──────────────────────────────────┤
│ 3 │ 1                                │
├───┼──────────────────────────────────┤
│ 4 │ null                             │
├───┼──────────────────────────────────┤
│ 5 │ undefined                        │
├───┼──────────────────────────────────┤
│ 6 │ Symbol(oh hello)                 │
├───┼──────────────────────────────────┤
│ 7 │ () => { }                        │
├───┼──────────────────────────────────┤
│ 8 │ ┌───┬──────────────────┐         │
│   │ │ 0 │ something        │         │
│   │ ├───┼──────────────────┤         │
│   │ │ 1 │ 0                │         │
│   │ ├───┼──────────────────┤         │
│   │ │ 2 │ false            │         │
│   │ ├───┼──────────────────┤         │
│   │ │ 3 │ 1                │         │
│   │ ├───┼──────────────────┤         │
│   │ │ 4 │ null             │         │
│   │ ├───┼──────────────────┤         │
│   │ │ 5 │ undefined        │         │
│   │ ├───┼──────────────────┤         │
│   │ │ 6 │ Symbol(oh hello) │         │
│   │ ├───┼──────────────────┤         │
│   │ │ 7 │ () => { }        │         │
│   │ └───┴──────────────────┘         │
├───┼──────────────────────────────────┤
│ 9 │ ┌───────────┬──────────────────┐ │
│   │ │ string    │ something        │ │
│   │ ├───────────┼──────────────────┤ │
│   │ │ number    │ 0                │ │
│   │ ├───────────┼──────────────────┤ │
│   │ │ boolean   │ false            │ │
│   │ ├───────────┼──────────────────┤ │
│   │ │ bigint    │ 1                │ │
│   │ ├───────────┼──────────────────┤ │
│   │ │ null      │ null             │ │
│   │ ├───────────┼──────────────────┤ │
│   │ │ undefined │ undefined        │ │
│   │ ├───────────┼──────────────────┤ │
│   │ │ symbol    │ Symbol(oh hello) │ │
│   │ ├───────────┼──────────────────┤ │
│   │ │ function  │ () => { }        │ │
│   │ └───────────┴──────────────────┘ │
└───┴──────────────────────────────────┘"
`;

exports[`stdio/printer printRecord should return an empty string for an empty array 1`] = `""`;
